builder.go

```diff
@@ -0,0 +1,247 @@
+package biscuit
+
+import (
+	"errors"
+	"io"
+
+	"github.com/flynn/biscuit-go/datalog"
+	"github.com/flynn/biscuit-go/pb"
+	"github.com/flynn/biscuit-go/sig"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	ErrDuplicateFact     = errors.New("biscuit: fact already exists")
+	ErrInvalidBlockIndex = errors.New("biscuit: invalid block index")
+)
+
+type Builder interface {
+	AddAuthorityFact(fact *Fact) error
+	AddAuthorityRule(rule *Rule) error
+	AddAuthorityCaveat(rule *Rule) error
+	AddRight(resource, right string) error
+	Build() (*Biscuit, error)
+}
+
+type builder struct {
+	rng  io.Reader
+	root sig.Keypair
+
+	symbolsStart int
+	symbols      *datalog.SymbolTable
+	facts        *datalog.FactSet
+	rules        []*datalog.Rule
+	caveats      []*datalog.Caveat
+	context      string
+}
+
+func NewBuilder(rng io.Reader, root sig.Keypair) Builder {
+	return BuilderWithSymbols(rng, root, DefaultSymbolTable)
+}
+
+func BuilderWithSymbols(rng io.Reader, root sig.Keypair, symbols *datalog.SymbolTable) Builder {
+	return &builder{
+		rng:  rng,
+		root: root,
+
+		symbolsStart: symbols.Len(),
+		symbols:      symbols,
+		facts:        new(datalog.FactSet),
+	}
+}
+
+func (b *builder) AddAuthorityFact(fact *Fact) error {
+	if len(fact.Predicate.IDs) == 0 {
+		fact.Predicate.IDs = []Atom{SymbolAuthority}
+	} else if fact.Predicate.IDs[0] != SymbolAuthority {
+		fact.Predicate.IDs[0] = SymbolAuthority
```

Comment:
I think we need to make a copy of the fact in these cases. In cases where the exact same "template" fact is used across multiple builders, we can end up modifying data being used in multiple places, causing races.